src/ripple/app/tx/impl/Payment.cpp

```diff
@@ -30,6 +30,19 @@ namespace ripple {
 
 // See https://ripple.com/wiki/Transaction_Format#Payment_.280.29
 
+XRPAmount
+Payment::calculateMaxSpend(STTx const& tx)
+{
+    auto const sendMax = tx[~sfSendMax];
+    auto const saDstAmount(tx.getFieldAmount(sfAmount));
+
+    /* If there's no sfSendMax in XRP, and the sfAmount isn't
+    in XRP, then the transaction can not send XRP. */
+    return sendMax ?
+        (sendMax->native() ? sendMax->xrp() : beast::zero) :
+        (saDstAmount.native() ? saDstAmount.xrp() : beast::zero);
+}
+
```

Comment:
Hm. `sendMax` can't safely be a reference because `STObject::operator[](OptionaledField<T> const&) const` doesn't return a reference type. But what I can do is
```
    if (tx.isFieldPresent(sfSendMax))
    {
        auto const& sendMax = tx[sfSendMax];
        return sendMax.native() ? sendMax.xrp() : beast::zero;
    }
```